### ISSUE TITLE
Behaviour of sol::optional<sol::object> has changed in the past few months

### ISSUE THREAD

**Issue by pvandommelen** [ThePhD/sol2#673]
I have a function exposed from cpp that looks something like this:
```
[](sol::optional<sol::object> optional_arg) {
    if (optional_arg) {
        ....
    }
}
```

In the version currently on master, the if branch is always entered. In the case where no argument is supplied, its value is nil. Valid, but unexpected for me.

In the version we were previously using (retrieved a while ago on 2017-12-12) it would not enter the if branch when no argument was supplied. 

Is this new behaviour intended?

**Comment by ThePhD** [comment]
I've been kicking this behavior around a lot. It should have been fixed in latest, but I guess not? Albeit, some people complained when I made it so `sol::object` wasn't a catch-all, even in the case of `none`. And to be fair, I sort of agree with them: `sol::object` is already optional on none-or-not: it's a "I accept everything" type that you're then supposed to apply discrimination upon to get what you want (with `.is<T>();`, `.get_type()` and friends).

I'm going to need to do a survey of this and #638 and other related issues...
